File: backend/api/languages.py
"""Languages API Routes

Provides language configuration and grammar data for frontend.
"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from core.logging import api_logger
from languages import get_module, list_languages

log = api_logger()

router = APIRouter()


# === Response Models ===

class CaseColorResponse(BaseModel):
    bg: str
    text: str
    border: str


class CaseResponse(BaseModel):
    id: str
    label: str
    hint: str
    color: CaseColorResponse


class GenderResponse(BaseModel):
    id: str
    label: str
    short: str


class NumberResponse(BaseModel):
    id: str
    label: str


class GrammarConfigResponse(BaseModel):
    cases: list[CaseResponse]
    genders: list[GenderResponse]
    numbers: list[NumberResponse]
    hasDeclension: bool
    hasConjugation: bool


class LanguageInfoResponse(BaseModel):
    code: str
    name: str
    nativeName: str


# === Endpoints ===

@router.get("/", response_model=list[LanguageInfoResponse])
async def get_available_languages():
    """Get list of available languages."""
    return list_languages()


@router.get("/{lang_code}", response_model=LanguageInfoResponse)
async def get_language_info(lang_code: str):
    """Get language info by code."""
    try:
        module = get_module(lang_code)
        return LanguageInfoResponse(code=module.code, name=module.name, nativeName=module.native_name)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{lang_code}/grammar", response_model=GrammarConfigResponse)
async def get_grammar_config(lang_code: str):
    """Get grammar configuration (cases, genders, numbers) for frontend."""
    try:
        module = get_module(lang_code)
        config = module.get_grammar_config()
        log.debug("grammar_config_fetched", language=lang_code)
        return config.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{lang_code}/declension-patterns")
async def get_declension_patterns(lang_code: str):
    """Get declension patterns for the language."""
    try:
        module = get_module(lang_code)
        patterns = module.get_declension_patterns()
        return {"patterns": patterns}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/services/languages.ts
/**
 * Languages Service
 * 
 * Handles language configuration API calls including grammar data.
 */
import { api, handleApiError } from '../lib/api';

// === Types ===

export interface CaseColor {
  bg: string;
  text: string;
  border: string;
}

export interface CaseConfig {
  id: string;
  label: string;
  hint: string;
  color: CaseColor;
}

export interface GenderConfig {
  id: string;
  label: string;
  short: string;
}

export interface NumberConfig {
  id: string;
  label: string;
}

export interface GrammarConfig {
  cases: CaseConfig[];
  genders: GenderConfig[];
  numbers: NumberConfig[];
  hasDeclension: boolean;
  hasConjugation: boolean;
}

export interface LanguageInfo {
  code: string;
  name: string;
  nativeName: string;
}

// Cache for grammar config
let grammarCache: Map<string, GrammarConfig> = new Map();

// === API Functions ===

/**
 * Get list of available languages
 */
export async function getAvailableLanguages(): Promise<LanguageInfo[]> {
  try {
    const response = await api.get<LanguageInfo[]>('/languages/');
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Get language info by code
 */
export async function getLanguageInfo(langCode: string): Promise<LanguageInfo> {
  try {
    const response = await api.get<LanguageInfo>(`/languages/${langCode}`);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Get grammar configuration for a language (cases, genders, numbers)
 */
export async function getGrammarConfig(langCode: string): Promise<GrammarConfig> {
  // Check cache first
  if (grammarCache.has(langCode)) {
    return grammarCache.get(langCode)!;
  }

  try {
    const response = await api.get<GrammarConfig>(`/languages/${langCode}/grammar`);
    grammarCache.set(langCode, response.data);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Clear grammar config cache
 */
export function clearGrammarCache(): void {
  grammarCache.clear();
}

// === Helper Functions ===

/**
 * Get case config by ID from grammar config
 */
export function getCaseById(grammar: GrammarConfig, caseId: string): CaseConfig | undefined {
  return grammar.cases.find(c => c.id === caseId);
}

/**
 * Get case label map from grammar config
 */
export function getCaseLabels(grammar: GrammarConfig): Record<string, string> {
  return grammar.cases.reduce((acc, c) => ({ ...acc, [c.id]: c.label }), {} as Record<string, string>);
}

/**
 * Get case hints map from grammar config
 */
export function getCaseHints(grammar: GrammarConfig): Record<string, string> {
  return grammar.cases.reduce((acc, c) => ({ ...acc, [c.id]: c.hint }), {} as Record<string, string>);
}

/**
 * Get case colors map from grammar config
 */
export function getCaseColors(grammar: GrammarConfig): Record<string, CaseColor> {
  return grammar.cases.reduce((acc, c) => ({ ...acc, [c.id]: c.color }), {} as Record<string, CaseColor>);
}
